group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java

```diff
@@ -10921,6 +10822,1048 @@ public void testLastConsumerProtocolMemberRebalanceTimeoutInConsumerGroup() {
+                    .setProtocolName("range"),
+                joinResult.joinFuture.get()
+            );
+            context.assertSyncTimeout(groupId, memberId1, request.rebalanceTimeoutMs());
+        }
+    }
+
+    @Test
+    public void testStableMemberJoiningConsumerGroupWithNewSubscriptionTransitionsToUnreleasedPartitions() throws Exception {
+        String groupId = "group-id";
+        Uuid fooTopicId = Uuid.randomUuid();
+        String fooTopicName = "foo";
+        Uuid barTopicId = Uuid.randomUuid();
+        String barTopicName = "bar";
+        Uuid zarTopicId = Uuid.randomUuid();
+        String zarTopicName = "zar";
+
+        for (short version = ConsumerProtocolSubscription.LOWEST_SUPPORTED_VERSION; version <= ConsumerProtocolSubscription.HIGHEST_SUPPORTED_VERSION; version++) {
+            String memberId1 = Uuid.randomUuid().toString();
+            String memberId2 = Uuid.randomUuid().toString();
+
+            MockPartitionAssignor assignor = new MockPartitionAssignor("range");
+            GroupMetadataManagerTestContext context = new GroupMetadataManagerTestContext.Builder()
+                .withAssignors(Collections.singletonList(assignor))
+                .withMetadataImage(new MetadataImageBuilder()
+                    .addTopic(fooTopicId, fooTopicName, 2)
+                    .addTopic(barTopicId, barTopicName, 1)
+                    .addTopic(zarTopicId, zarTopicName, 1)
+                    .addRacks()
+                    .build())
+                .withConsumerGroup(new ConsumerGroupBuilder(groupId, 10)
+                    .withSubscriptionMetadata(new HashMap<String, TopicMetadata>() {
+                        {
+                            put(fooTopicName, new TopicMetadata(fooTopicId, fooTopicName, 2, mkMapOfPartitionRacks(2)));
+                            put(barTopicName, new TopicMetadata(barTopicId, barTopicName, 1, mkMapOfPartitionRacks(1)));
+                        }
+                    })
+                    .withMember(new ConsumerGroupMember.Builder(memberId1)
+                        .setState(MemberState.STABLE)
+                        .setMemberEpoch(10)
+                        .setPreviousMemberEpoch(10)
+                        .setRebalanceTimeoutMs(500)
+                        .setClientId("client")
+                        .setClientHost("localhost/127.0.0.1")
+                        .setSubscribedTopicNames(Arrays.asList(fooTopicName, barTopicName))
+                        .setSupportedClassicProtocols(GroupMetadataManagerTestContext.toRangeProtocol(
+                            Arrays.asList(fooTopicName, barTopicName),
+                            Arrays.asList(new TopicPartition(fooTopicName, 0), new TopicPartition(fooTopicName, 1)),
+                            version))
+                        .setPartitionsPendingRevocation(mkAssignment(
+                            mkTopicAssignment(fooTopicId, 1)))
```

Comment:
This does not seem correct. If the member is Stable, he cannot have unrevoked partitions too.